rust-runtime/aws-smithy-experimental/src/hyper_1_0.rs

```diff
@@ -0,0 +1,1223 @@
+        HttpConnectorFuture::new(async move {
+            let response = fut
+                .await
+                .map_err(downcast_error)?
+                .map(SdkBody::from_body_1_x);
+            match HttpResponse::try_from(response) {
+                Ok(response) => Ok(response),
+                Err(err) => Err(ConnectorError::other(err.into(), None)),
+            }
+        })
+    }
+}
+
+/// Downcast errors coming out of hyper into an appropriate `ConnectorError`
+fn downcast_error(err: BoxError) -> ConnectorError {
+    // is a `TimedOutError` (from aws_smithy_async::timeout) in the chain? if it is, this is a timeout
+    if find_source::<TimedOutError>(err.as_ref()).is_some() {
+        return ConnectorError::timeout(err);
+    }
+    // is the top of chain error actually already a `ConnectorError`? return that directly
+    let err = match err.downcast::<ConnectorError>() {
+        Ok(connector_error) => return *connector_error,
+        Err(box_error) => box_error,
+    };
+    // generally, the top of chain will probably be a hyper error. Go through a set of hyper specific
+    // error classifications
+    let err = match find_source::<hyper::Error>(err.as_ref()) {
+        Some(hyper_error) => return to_connector_error(hyper_error)(err),
+        None => err,
+    };
+
+    // otherwise, we have no idea!
+    ConnectorError::other(err, None)
+}
+
+/// Convert a [`hyper::Error`] into a [`ConnectorError`]
+fn to_connector_error(err: &hyper::Error) -> fn(BoxError) -> ConnectorError {
+    if err.is_timeout() || find_source::<HttpTimeoutError>(err).is_some() {
+        return ConnectorError::timeout;
+    }
+    if err.is_user() {
+        return ConnectorError::user;
+    }
+    if err.is_closed() || err.is_canceled() || find_source::<std::io::Error>(err).is_some() {
+        return ConnectorError::io;
+    }
+    // We sometimes receive this from S3: hyper::Error(IncompleteMessage)
+    if err.is_incomplete_message() {
+        return |err: BoxError| ConnectorError::other(err.into(), Some(ErrorKind::TransientError));
+    }
+
+    if let Some(h2_err) = find_source::<h2::Error>(err) {
+        if h2_err.is_go_away()
+            || (h2_err.is_reset() && h2_err.reason() == Some(Reason::REFUSED_STREAM))
+        {
+            return ConnectorError::io;
+        }
+    }
+
+    tracing::warn!(err = %DisplayErrorContext(&err), "unrecognized error from Hyper. If this error should be retried, please file an issue.");
+    |err: BoxError| ConnectorError::other(err.into(), None)
+}
+
+fn find_source<'a, E: Error + 'static>(err: &'a (dyn Error + 'static)) -> Option<&'a E> {
+    let mut next = Some(err);
+    while let Some(err) = next {
+        if let Some(matching_err) = err.downcast_ref::<E>() {
+            return Some(matching_err);
+        }
+        next = err.source();
+    }
+    None
+}
+
+// TODO(https://github.com/awslabs/aws-sdk-rust/issues/1090): CacheKey must also include ptr equality to RuntimeComponents
```

Comment:
I think pointer equality will end up ballooning the cached connectors into a memory leak if someone uses a config override on a operation. We'll have to come up with something else.

Reply:
ah yeah—that comment is badly worded. It should be the specific components of RC that are used.